presto-benchmark-driver/src/main/java/com/facebook/presto/benchmark/driver/BenchmarkQueryRunner.java

```diff
@@ -149,53 +152,56 @@ public BenchmarkQueryResult execute(Suite suite, ClientSession session, Benchmar
-                if (data != null) {
-                    for (List<Object> objects : data) {
-                        schemas.add(objects.get(0).toString());
-                    }
-                }
-            }
-
-            // verify final state
-            if (client.isClosed()) {
-                throw new IllegalStateException("Query aborted by user");
-            }
-
-            if (client.isGone()) {
-                throw new IllegalStateException("Query is gone (server restarted?)");
+            AtomicBoolean success = new AtomicBoolean(true);
+            execute(
+                    session,
+                    "show schemas",
+                    queryData -> {
+                        for (List<Object> objects : queryData.getData()) {
+                            schemas.add(objects.get(0).toString());
+                        }
+                    },
+                    queryError -> {
+                        success.set(false);
+                        handleFailure(getCause(queryError));
+                    });
+            if (success.get()) {
+                return schemas.build();
             }
+        }
+    }
 
-            QueryError resultsError = client.finalStatusInfo().getError();
-            if (resultsError != null) {
-                RuntimeException cause = null;
-                if (resultsError.getFailureInfo() != null) {
-                    cause = resultsError.getFailureInfo().toException();
-                }
-                handleFailure(cause);
-
-                continue;
-            }
+    private StatementStats execute(ClientSession session, String name, String query)
+    {
+        return execute(
+                session,
+                query,
+                queryData -> {}, // we do not process the output
+                resultsError -> {
+                    throw new BenchmarkDriverExecutionException(format("Query %s failed: %s", name, resultsError.getMessage()), getCause(resultsError));
+                });
+    }
 
-            return schemas.build();
+    private static RuntimeException getCause(QueryError queryError)
+    {
+        if (queryError.getFailureInfo() != null) {
+            return queryError.getFailureInfo().toException();
         }
+        return null;
     }
 
-    private StatementStats execute(ClientSession session, String name, String query)
+    private StatementStats execute(ClientSession session, String query, Consumer<QueryData> queryDataConsumer, Consumer<QueryError> queryErrorConsumer)
     {
         // start query
         StatementClient client = new StatementClient(okHttpClient, session, query);
 
         // read query output
         while (client.isValid() && client.advance()) {
-            // we do not process the output
+            QueryData queryData = client.currentData();
+            if (queryData != null) {
+                queryDataConsumer.accept(queryData);
```

Comment:
shouldn't consumer be notified if data is null? What does it mean actually? That error happened. Or also an empty result set.

Are queryData==null and resutlError==null mutually exclusive?

Reply:
AFAIK, `client.currentData()` is not-null.

Reply:
However, `client.currentData().getData()` is nullable.

Reply:
> Are queryData==null and resutlError==null mutually exclusive?

Practically it looks so, but I couldn't find any enforcement of this.